hazelcast/src/main/java/com/hazelcast/query/impl/predicates/AbstractPredicate.java

```diff
@@ -89,7 +90,15 @@ private boolean convertAndApplyForSingleAttributeValue(Object attributeValue) {
             }
             attributeValue = AbstractJsonGetter.convertFromJsonValue((JsonValue) attributeValue);
         }
-        return applyForSingleAttributeValue((Comparable) attributeValue);
+        if (attributeValue instanceof Comparable) {
+            return applyForSingleAttributeValue((Comparable) attributeValue);
+        }
+        if (attributeValue instanceof GenericRecord) {
+            throw new QueryException(attributeName + " field can not be compared, because "
+                    + "the user class could not be constructed. GenericRecord representation " + attributeValue);
```

Comment:
Error messages shouldn't contain user data.

Reply:
Is this because of security reasons? I have never followed this rule. I am just curious about the reasoning.

Reply:
I have replaced the data with metadata. ClassDefinition or the class name.

Reply:
Yes, it's for security. Error messages can appear anywhere in logs, in various notifications users might set up, in the users UI etc. They can also appear to users not owning the data: the engine might be scanning rows not belonging to the current user and encounter errors there - those data should not appear to the user. It's a general rule - don't include user data in error messages.